src/optimizer/cardinality_estimator.cpp

```diff
@@ -0,0 +1,477 @@
+
+			// sometimes base stats is null (test_709.test) returns null for base stats while
+			// there is still a catalog table. Anybody know anything about this?
+			auto base_stats = catalog_table->storage->GetStatistics(context, actual_binding.column_index);
+			if (base_stats) {
+				count = base_stats->GetDistinctCount();
+			}
+
+			// means you have a direct filter on a column. The count/total domain for the column
+			// should be decreased to match the predicted total domain matching the filter.
+			// We decrease the total domain for all columns in the equivalence set because filter pushdown
+			// will mean all columns are affected.
+			if (direct_filter) {
+				count = node->GetCardinality();
+			}
+
+			// HLL has estimation error, count can't be greater than cardinality of the table before filters
+			if (count > node->GetBaseTableCardinality()) {
+				count = node->GetBaseTableCardinality();
+			}
+		} else {
+			// No HLL. So if we know there is a direct filter, reduce count to cardinality with filter
+			// otherwise assume the total domain is still the cardinality
+			if (direct_filter) {
+				count = node->GetCardinality();
+			} else {
+				count = node->GetBaseTableCardinality();
+			}
+		}
+
+		for (idx_t ind = 0; ind < equivalent_relations.size(); ind++) {
+			column_binding_set_t i_set = equivalent_relations.at(ind);
+			if (i_set.count(key) == 1) {
+				if (catalog_table) {
+					if (equivalent_relations_tdom_hll.at(ind) < count) {
+						equivalent_relations_tdom_hll.at(ind) = count;
+					}
+					if (equivalent_relations_tdom_no_hll.at(ind) > count) {
+						equivalent_relations_tdom_no_hll.at(ind) = count;
+					}
+				} else {
+					if (equivalent_relations_tdom_no_hll.at(ind) > count) {
+						equivalent_relations_tdom_no_hll.at(ind) = count;
+					}
+					// If we join a parquet/csv table with a catalog table that has HLL
+					// then count = cardinality.
+					// In this case, prefer the lower values.
+					// equivalent_relations_tdom_hll is initialized to 0 for all equivalence relations
+					// so check for 0 to make sure an ideal value exists.
+					if (equivalent_relations_tdom_hll.at(ind) > count || equivalent_relations_tdom_hll.at(ind) == 0) {
+						equivalent_relations_tdom_hll.at(ind) = count;
+					}
+				}
+				break;
+			}
+		}
+	}
+}
+
+TableFilterSet *CardinalityEstimator::GetTableFilters(LogicalOperator *op) {
+	// First check table filters
+	auto get = GetLogicalGet(op);
+	if (get) {
+		return &get->table_filters;
+	}
+	return nullptr;
+}
+
+idx_t CardinalityEstimator::InspectConjunctionAND(idx_t cardinality, idx_t column_index, ConjunctionAndFilter *filter,
+                                                  TableCatalogEntry *catalog_table) {
+	auto has_equality_filter = false;
+	auto cardinality_after_filters = cardinality;
+	for (auto &child_filter : filter->child_filters) {
+		if (child_filter->filter_type == TableFilterType::CONSTANT_COMPARISON) {
```

Comment:
We prefer early-out to avoid deep indentation when possible, e.g.:

```cpp
if (child_filter->filter_type != TableFilterType::CONSTANT_COMPARISON) {
    continue;
}
// ...
```